Fix interactions between macros and by-name implicits

milessabin opened the pull request:
While prototyping a version of shapeless which replaces `Lazy` with by-name implicits it emerged that implicit resolutions which involve both by-name implicits and implicit macros can produce dictionaries
which type check but which have broken owner chains.

This had been anticipated (see 5dd99093) but the ownership change performed there wasn't thoroughgoing enough.

In addition, an unnecessary check in `Implicits.scala`, excluding implicit expansion sites with empty trees when attempting to determine if an implicit resolution was recursive, meant that implicit macros, which
always have empty trees in this case, were never detected as recursive.

Finally, the implicit dictionary is now encoded as,
```scala
final class LazyDefns$n extends java.io.Serializable {
  final val rec$m ...
}
val lazyDefns$n = new LazyDefns$
lazyDefns$n.rec$m
```
rather than as an object, to eliminate the unecessary overhead of object lazy initialization and to allow serialzation in rare cases where the enclosing value is captured.

With this commit a `Lazy`-free shapeless build is possible.

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Contexts.scala:
```diff
@@ -339,6 +339,30 @@ trait Contexts { self: Analyzer =>
       else {
         val typer = newTyper(this)
 
+        def changeNonLocalOwners(tree: Tree, newowner: Symbol): Unit = {
```
Could you move this closer to `class ChangeOwnerTraverser`, so that it's more discoverable for potential future usage?

adriaanm replied:
Maybe, after some refactoring, this could be a subclass of ChangeOwnerTraverser?

milessabin replied:
I've moved this next to `ChangeOwnerTraverser` and lifted the `LocalOwnerTraverser` out. It doesn't really work as a subclass of the former.

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Contexts.scala:
```diff
@@ -351,35 +375,56 @@ trait Contexts { self: Analyzer =>
-          }
+          val vname = newTermName(typer.fresh.newName("lazyDefns$"))
+          val vdef0 =
+            ValDef(Modifiers(FINAL | SYNTHETIC), vname, Ident(cname), New(Ident(cname), Nil))
```
Could include a signature for the valdef. It's simply `cdef.symbol.tpe_*`, right?

milessabin replied:
What do you mean by that? The `ValDef` has a tpt: `Ident(cname)`.

milessabin replied:
Ahh ... OK, I think I get it :-)

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Contexts.scala:
```diff
@@ -351,35 +375,56 @@ trait Contexts { self: Analyzer =>
+
+          val cname = newTypeName(typer.fresh.newName("LazyDefns$"))
+          val cdef0 =
+            ClassDef(Modifiers(FINAL | SYNTHETIC), cname, Nil,
```
If possible, it would be preferable to use an approach closer to the one in synthesizePartialFunction (https://github.com/scala/scala/blob/036143bad49e07bf131386bf2df39ec889e24dc9/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L2805-L2814), and call the ClassDef factory method at https://github.com/scala/scala/blob/80f165a631c675eeb593031f8cff541c7e8ca15e/src/compiler/scala/tools/nsc/ast/Trees.scala#L61

adriaanm reviewed (dismissed):
Let's see if we can refine this a bit. Could you also split out the different fixes into separate commits?

milessabin committed:
Don't exclude macros when checking for implicit recursion

An unnecessary check in Implicits.scala excludes implicit expansion
sites with empty trees when attempting to determine if an implicit
resolution was recursive meant that implicit macros, which always have
empty trees in this case, were never detected as recursive.

milessabin committed:
Fix interactions between macros and by-name implicits

While prototyping a version of shapeless which replaces Lazy with
by-name implicits it emerged that implicit resolutions which involve
both by-name implicits and implicit macros can produce dictionaries
which type checked but which had broken owner chains.

This had been anticipated (see 5dd99093) but the ownership change
performed there wasn't thoroughgoing enough.

The implicit dictionary is now encoded as,

  final class LazyDefns$n extends java.io.Serializable {
    final val rec$m ...
  }
  val lazyDefns$n = new LazyDefns$
  lazyDefns$n.rec$m

rather than as an object, to eliminate the unecessary overhead of object
lazy initialization and to allow serialzation in rare cases where the
enclosing value is captured.

With this commit a Lazy-free shapeless build is possible.

milessabin commented:
Also rebased.

milessabin commented:
The failure on Travis appears to be another instance of https://github.com/scala/scala-dev/issues/570.

SethTisue commented:
@milessabin rebasing should fix Travis

milessabin commented:
Rebased.

milessabin commented:
Travis failure this time is a gateway timeout ... I've restarted the job.

milessabin commented:
@adriaanm this is ready to go now ...

lrytz commented:
This is the same SOE as initially reported in https://github.com/scala/scala-dev/issues/570, if you look a little bit further down:

```
[error] java.lang.StackOverflowError
[error] 	at scala.tools.nsc.typechecker.Typers$Typer.makeAccessible(Typers.scala:548)
[error] 	at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$60(Typers.scala:5074)
[error] 	at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$silent$2(Typers.scala:717)
[error] 	at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:717)
[error] 	at scala.tools.nsc.typechecker.Typers$Typer.typedSelect$1(Typers.scala:5074)
[error] 	at scala.tools.nsc.typechecker.Typers$Typer.typedSelectOrSuperCall$1(Typers.scala:5131)
[error] 	at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5676)
[error] 	at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:797)
```

I guess someone catches the SOE.

milessabin commented:
@lrytz am I right that the SOE isn't an issue for this PR, it's still https://github.com/scala/scala-dev/issues/570?

lrytz commented:
yes!

milessabin commented:
Rebased following https://github.com/scala/scala/pull/7375.

milessabin commented:
@SethTisue The failure I'm seeing here is a gateway timeout: https://travis-ci.org/scala/scala/jobs/452899122#L5233.

milessabin commented:
Rebased.

adriaanm commented:
Sorry about the delay.

milessabin commented:
Thanks!
